emacs/config.org
#+TITLE: Emacs Config
#+AUTHOR: soham
#+OPTIONS: toc:t date:t

This is my Emacs configuration. Based on [[https://github.com/jamiecollinson/dotfiles/blob/master/config.org/][Jamie Collision's]] and [[https:github.com/DiegoVicen/my-emacs/blob/master/README.org][Diego Vicente's]]
configurations, with some additions from [[https:pragamaticemacs.com][Pragmatic Emacs]].

* Basic Setup
** Configure package sources

Add repositories from which Emacs will load packages. At present I am using the
MELPA repository exclusively. MELPA Stable might be a better choice for
stability,  but it does not have all the packages that I use.

#+BEGIN_SRC emacs-lisp
(when (>= emacs-major-version 24)
  (require 'package)
  (add-to-list 'package-archives
	       '("melpa" . "https://melpa.org/packages/") t)
  (add-to-list 'package-archives
               '("org" . "http://orgmode.org/elpa/") t))
#+END_SRC

*** TODO Check if MELPA Stable has all the required packages.

** Increase garbage collector threshold

The default garbage collection threshold is 800kB, increasing this to 10MB for
startup increases speed (from 11.0s -> 9.7s when I tested).

#+BEGIN_SRC emacs-lisp
(setq gc-cons-threshold 10000000)

;; Restore after startup
(add-hook 'after-init-hook
          (lambda ()
            (setq gc-cons-threshold 1000000)
            (message "gc-cons-threshold restored to %S"
                     gc-cons-threshold)))
#+END_SRC

** Configuration directory

Directory for all my configurations.

#+BEGIN_SRC emacs-lisp
(setq configuration-dir "~/dots/emacs/")
#+END_SRC

** Edit and reload config

These are keybindings to easily edit and reload this configuration file.

#+BEGIN_SRC emacs-lisp
(defun reload-emacs-configuration()
  "Reload the configuration"
  (interactive)
  (load "~/.emacs"))

(defun open-emacs-configuration ()
  "Open the configuration.org file in buffer"
  (interactive)
  (find-file (concat configuration-dir "config.org")))

(global-set-key (kbd "C-c R") 'reload-emacs-configuration)
(global-set-key (kbd "C-c I") 'open-emacs-configuration)
#+END_SRC

** Backups folder

I absolutely hate the =filename#= spamming that Emacs does. At the same time I
love that I can have a backup in case I accidentally powered off my computer
without saving the files I was working on. This setup centralises the backup
files created by Emacs.

#+BEGIN_SRC emacs-lisp
(setq backup-directory-alist '(("." . "~/.emacs.d/backup"))
      backup-by-copying t    ; Don't delink hardlinks
      version-control t      ; Use version numbers on backups
      delete-old-versions t  ; Automatically delete excess backups
      kept-new-versions 20   ; how many of the newest versions to keep
      kept-old-versions 5    ; and how many of the old
      )
#+END_SRC

** Set custom settings to load in own file

This stops Emacs adding customised settings to =init.el=. I try to avoid using
customize anyway, preferring programmatic control of variables. Creating it as a
temporary file effectively disables it (i.e. any changes are session local).

#+BEGIN_SRC emacs-lisp
(setq custom-file (make-temp-file "emacs-custom"))
#+END_SRC

** Add custom packages to load path

By default Emacs only includes files directly under =user-emacs-directory=
(usually =~/.emacs.d/=), so we need to add any folders containing custom
packages. I put my scripts under =lisp= directory in my configuration directory
and symlink with =stow=.

#+BEGIN_SRC emacs-lisp
(add-to-list 'load-path (concat configuration-dir "lisp"))
#+END_SRC

** Shorter yes or no prompts

Single key strokes are probably better than triple key strokes.

#+BEGIN_SRC emacs-lisp
(fset 'yes-or-no-p 'y-or-n-p)
#+END_SRC

** Disable default startup screen

I have no particular use for the default startup screen. I prefer the dashboard
used by Spacemacs / Doom Emacs, which we will come to later. For now let us
disable the startup-screen. While we are at it also start the =scratch= buffer
in =org-mode= and disable the default =scratch= buffer message.

#+BEGIN_SRC emacs-lisp
(setq inhibit-startup-screen t
      initial-scratch-message ""
      initial-major-mode 'org-mode)
#+END_SRC

** Disable the bell

The bell is a absolute no-no.

#+BEGIN_SRC emacs-lisp
(setq ring-bell-function 'ignore)
#+END_SRC

** Emacs abbreviations

I have not used Emacs abbreviations much, but seems like they can be
useful, probably when I am typing my thesis. Just in case let us fix some
choices for the abbreviations.

#+BEGIN_SRC emacs-lisp
(setq save-abbrevs 'silent
      abbrev-file-name "~/.emacs.d/abbrev_defs")
#+END_SRC

** Tabs vs Spaces

Going to go with spaces on this one. I will also fix tab width to 2.

#+BEGIN_SRC emacs-lisp
(setq-default indent-tabs-mode nil)
(setq tab-width 2)
#+END_SRC

** Fix scroll

Smoother scroll whether I am using the keyboard or the mouse (sometimes the
mouse can be useful).

#+BEGIN_SRC emacs-lisp
(setq scroll-step            1
      scroll-conservatively  10000
      mouse-wheel-scroll-amount '(1 ((shift) . 1))
      mouse-wheel-progressive-speed nil
      mouse-wheel-follow-mouse 't)
#+END_SRC

** Bootstrap use-package

I am using =use-package= to install all the other Emacs packages that I
use. There are quite a few excellent Emacs package managers on the market. But
this is the one I am used to, and I would rather just install the packages and
get to work that figure out how to use a different package manager. This
installs =use-package=, if it is not already installed.

#+BEGIN_SRC emacs-lisp
(package-initialize)
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))
#+END_SRC

By default all packages should be installed from package manager as that's the
usual path. This is equivalent to setting =:ensure t= on each call to
=use-package=. To disable we just have to set =:ensure nil= (this is done
automatically for any packages using =:load-path= so shouldn't generally be
needed).

#+BEGIN_SRC emacs-lisp
(setq use-package-always-ensure t)
#+END_SRC

** Automatic package updates

No one wants to check for package updates manually, unless they are concerned
about breakages. I am not for the most part, so I will use =auto-package-update=.

#+BEGIN_SRC emacs-lisp
(use-package auto-package-update
  :config
  (setq auto-package-update-delete-old-versions t)
  (setq auto-package-update-hide-results t)
  (setq auto-package-update-prompt-before-update t)
  (auto-package-update-maybe))
#+END_SRC

** Benchmark startup

I want to check how much time Emacs takes to startup, so that I can optimize it
if required. There's a nifty package called =benchmark-init= that records
startup time by package. It only records things after it is initialised, so it
needs to be put as early in config as possible.

#+BEGIN_SRC emacs-lisp
(use-package benchmark-init
  :hook
  ;; To disable collection of benchmark data after init is done.
  (after-init . benchmark-init/deactivate)
  (after-init . (lambda () (message "loaded in %s" (emacs-init-time)))))
#+END_SRC

** Environment management

Since I usually use the Emacs GUI the =exec-path= variable is left empty at
startup. We need to load the contents of =$PATH= to the =exec-path= variable and
call it at start.

#+BEGIN_SRC emacs-lisp
(use-package exec-path-from-shell
  :demand t
  :config
  (setq exec-path-from-shell-variables
        '("PATH" "MANPATH" "DESKTOP_SESSION"))
  (exec-path-from-shell-initialize))
#+END_SRC

** Insert new line without breaking

Vim has this nifty shortcut =o=, which inserts a new line below the line in
which the cursor is. I do not want to go full =evil=, but I do like this one
feature of vim. To have the same behavior in Emacs, I found this custom
function that I bound to C-o.

#+BEGIN_SRC emacs-lisp
(defun insert-new-line-below ()
  "Add a new line below the current line"
  (interactive)
  (let ((oldpos (point)))
    (end-of-line)
    (newline-and-indent)))

(global-set-key (kbd "C-o") 'insert-new-line-below)
#+END_SRC

** Move buffers around

If we want to swap buffers location in frames, there’s no fast way to do it in
Emacs by default. To do it, a good option that I found is to use buffer-move
package, and use these key bindings.

#+BEGIN_SRC emacs-lisp
(use-package buffer-move
  :bind (("C-c w <up>"    . buf-move-up)
         ("C-c w <down>"  . buf-move-down)
         ("C-c w <left>"  . buf-move-left)
         ("C-c w <right>" . buf-move-right)))
#+END_SRC

** Redefining sentences

For some reason Emacs assumes that we end sentences with a period and /two/
whitespaces. I generally use a single whitespace after the period. To move in
sentences we need to redefine sentence endings for Emacs.

#+BEGIN_SRC emacs-lisp
(setq-default sentence-end-double-space nil)
#+END_SRC

** Fill-column

I prefer code and text to respect the 80 characters per line limit, even though
sometimes, particularly with code, it is not possible at all.

#+BEGIN_SRC emacs-lisp
(setq-default fill-column 80)
#+END_SRC

** Auto-fill comments

For our comments (only comments, not code) to be automatically filled in
programming modes, we can use this function:

#+BEGIN_SRC emacs-lisp
(defun comment-auto-fill ()
  (setq-local comment-auto-fill-only-comments t)
  (auto-fill-mode 1))

(add-hook 'prog-mode-hook 'comment-auto-fill)
#+END_SRC

** Delete selection

Emacs by default does not follow the contemporary behavior of allowing
highlighted text to be deleted. But that is whad I am more used to. So we will
activate the =delete-selection-mode=.

#+BEGIN_SRC emacs-lisp
(delete-selection-mode 1)
#+END_SRC

** Scroll in the compilation buffer

Automatically scroll the compilation buffer as the output is printed.

#+BEGIN_SRC emacs-lisp
(setq compilation-scroll-output t)
#+END_SRC

** Delete trailing whitespaces

I *never* want whitespace at the end of lines. Remove it on save.

#+BEGIN_SRC emacs-lisp
(add-hook 'before-save-hook 'delete-trailing-whitespace)
#+END_SRC

** Highlight the current line.

This highlights the line the cursor is on. Helps me to focus.

#+BEGIN_SRC emacs-lisp
(global-hl-line-mode 1)
#+END_SRC

** Set the fringe as the background

In the GUI mode, each Emacs window has narrow fringes on the left and right
edges. This function allows to set the fringe color the same as the background,
which makes it look flatter and more minimalist.

#+BEGIN_SRC emacs-lisp
(defun set-fringe-as-background ()
  "Force the fringe to have the same color as the background"
  (set-face-attribute 'fringe nil
                      :foreground (face-foreground 'default)
                      :background (face-background 'default)))
#+END_SRC

* Graphical Interface
** Disable GUI defaults

I prefer a minimalistic looks which means that the GUI defaults have to take the
highway.

#+BEGIN_SRC emacs-lisp
(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)
#+END_SRC

** Fonts!!!

Who does not like a good font? My choice is Adobe's Source Code Pro.

#+BEGIN_SRC emacs-lisp
(set-frame-font "Source Code Pro 12" nil t)
#+END_SRC

** Setting the theme

=doom-themes= is an excellent collection of themes. I sometimes cycle through
them, using =counsel-load-theme= to load the themes. For now I will set
=doom-solarized-light= as the default theme.

#+BEGIN_SRC emacs-lisp
(use-package doom-themes
  :config
  (load-theme 'doom-solarized-light t))
#+END_SRC

** Icons!!!

=all-the-icons= has all the icons! So let us use it.

#+BEGIN_SRC emacs-lisp
(use-package all-the-icons
  :after font-lock+)
#+END_SRC

** Modeline

I hardly use the mouse with Emacs, and I have disabled all default mouse
features. But I am going to make an exception for the =moody= + =minions=
alternative to the builtin modeline.

#+BEGIN_SRC emacs-lisp
(use-package minions
  :config
  (setq minions-mode-line-lighter "[+]")
  (minions-mode))

(use-package moody
  :config
  (moody-replace-mode-line-buffer-identification)
  (moody-replace-vc-mode)
  (setq-default x-underline-at-descent-line t
                column-number-mode t))
#+END_SRC

* Packages \ Tools
** Crux

A Collection of Ridiculously Useful eXtensions for Emacs. =crux= bundles a few
useful interactive commands to enhance your overall Emacs experience. I have the
following settings:
- =C-a= Move to the first non-whitespace character on a line, and then to
  between that and the beginning of the line
- =C-k= First kill to end of line, then kill the whole line
- =C-c o= Open the currently visited file with an external program
- =C-S-RET= Insert an empty line above the current line and indent it properly
- =S-RET= Insert an empty line and indent it properly
- =C-c n= Fix indentation in buffer and strip whitespace
- =C-c f= Open recently visited file
- =C-c D= Delete current file and buffer
- =C-c e= Eval a bit of Emacs Lisp code and replace it with its result
- =C-c d= Duplicate the current line (or region)
- =C-c TAB= Indent and copy region to clipboard

#+BEGIN_SRC emacs-lisp
(use-package crux
  :demand
  :bind
  (("C-a" . crux-move-beginning-of-line)
   ("C-k" . crux-smart-kill-line)
   ("C-c o" . crux-open-with)
   ("C-S-<return>" . crux-smart-open-line-above)
   ("S-<return>" . crux-smart-open-line)
   ("C-c n" . crux-cleanup-buffer-or-region)
   ("C-c f" . crux-recentf-find-file)
   ("C-c D" . crux-delete-file-and-buffer)
   ("C-c e" . crux-eval-and-replace)
   ("C-c d" . crux-duplicate-current-line-or-region)
   ("C-c <tab>" . crux-indent-rigidly-and-copy-to-clipboard)))
#+END_SRC

** Dash

Modern list api for Emacs.

#+BEGIN_SRC emacs-lisp
(use-package dash)
(use-package dash-functional)
#+END_SRC
** Dashboard

This is the dashboard that is used in Spacemacs / Doom Emacs. I find it quite
useful, though sometimes it does not update the recent files as early as I
prefer.

#+BEGIN_SRC emacs-lisp
(use-package dashboard
  :config
  (dashboard-setup-startup-hook))
#+END_SRC

** Dired enhancements

Emac's =dired= is probably the best file manager that I have used. A close
second is =ranger=. Here are some packages that add some of the functionalities
of =ranger= that I particularly like to =dired=.

#+BEGIN_SRC emacs-lisp
  (use-package dired-hacks-utils)
  (use-package dired-ranger
    :after (dash)
    :bind (:map dired-mode-map
                ("W" . dired-ranger-copy)
                ("X" . dired-ranger-move)
                ("Y" . dired-ranger-paste)))
  (use-package dired-filter
    :after (dash)
    :bind (:map dired-mode-map
                ("F" . dired-filter-map)))
  (use-package dired-avfs)
  (use-package dired-rainbow)
  (use-package dired-collapse)
  (use-package dired-narrow
    :bind (:map dired-mode-map
                ("/" . dired-narrow)))
  (use-package dired-subtree
    :bind (:map dired-mode-map
                ("<tab>" . dired-subtree-toggle)
                ("<backtab>" . dired-subtree-cycle)))
#+END_SRC

There is an =all-the-icons= package for =dired=. So let us add that too!

#+BEGIN_SRC emacs-lisp
(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))
#+END_SRC

Default switches for =dired=.

#+BEGIN_SRC emacs-lisp
(setq dired-listing-switches "-lath --group-directories-first")
#+END_SRC

** Command completion

=smart M-x= suggests =M-x= commands based on recency and frequency. I don't tend
to use it directly but =counsel= uses it to order suggestions.

#+BEGIN_SRC emacs-lisp
(use-package smex)
#+END_SRC

=ivy= is a generic completion framework which uses the minibuffer. Turning on
=ivy-mode= enables replacement of lots of built in =ido=
functionality. =ivy-rich= adds a friendlier interface to =ivy=.

#+BEGIN_SRC emacs-lisp
(use-package ivy
  :diminish ivy-mode
  :config
  (ivy-mode t)
  (setq ivy-display-style 'fancy))

(use-package ivy-rich
  :after ivy
  :config
  ;; Disable TRAMP buffers extended information to prevent slowdown
  (setq ivy-rich-parse-remote-buffer nil)
  :init (ivy-rich-mode 1))

(use-package all-the-icons-ivy-rich
  :after ivy-rich
  :init (all-the-icons-ivy-rich-mode 1))

(setcdr (assq t ivy-format-functions-alist) #'ivy-format-function-line)
#+END_SRC

By default =ivy= starts filters with =^=. I don't normally want that and can
easily type it manually when I do.

#+BEGIN_SRC emacs-lisp
(setq ivy-initial-inputs-alist nil)
#+END_SRC

=counsel= is a collection of =ivy= enhanced versions of common Emacs commands. I
haven't bound much as =ivy-mode= takes care of most things.

#+BEGIN_SRC emacs-lisp
(use-package counsel
  :bind (("M-x" . counsel-M-x)))
#+END_SRC

=swiper= is an =ivy= enhanced version of isearch.

#+BEGIN_SRC emacs-lisp
(use-package swiper)
(global-set-key (kbd "C-s") 'counsel-grep-or-swiper)
;; (global-set-key (kbd "M-s") 'isearch-forward)
;; (global-set-key (kbd "M-r") 'isearch-backward)
#+END_SRC

=hydra= presents menus for =ivy= commands.

#+BEGIN_SRC emacs-lisp
(use-package ivy-hydra)
#+END_SRC

** Suggest next key

Suggest next keys based on currently entered key combination.

#+BEGIN_SRC emacs-lisp
(use-package which-key
  :diminish which-key-mode
  :hook
  (after-init . which-key-mode))
#+END_SRC

** Better undo

=undo-tree= visualises undo history as a tree for easy navigation.

#+BEGIN_SRC emacs-lisp
(use-package undo-tree
  :defer 5
  :diminish global-undo-tree-mode
  :config
  (global-undo-tree-mode 1))
#+END_SRC

** Navigation

One of the most important features of an advanced editor is quick text
navigation. =avy= lets us jump to any character or line quickly.

#+BEGIN_SRC emacs-lisp
(use-package avy
  :bind (("M-s" . avy-goto-word-1)))
#+END_SRC

=ace-window= lets us navigate between windows in the same way as =avy=. Once
activated it has useful sub-modes like =x= to switch into window deletion mode.

#+BEGIN_SRC emacs-lisp
(use-package ace-window
  :config
  (global-set-key (kbd "C-x o") 'ace-window)
  (setq aw-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l)))
#+END_SRC

** Easier selection

=expand-region= expands the region around the cursor semantically depending on
mode. Hard to describe but a killer feature.

#+BEGIN_SRC emacs-lisp
(use-package expand-region
  :bind ("C-=" . er/expand-region))
#+END_SRC

** Emoji support.

This is useful when working with html.

#+BEGIN_SRC emacs-lisp
(use-package emojify)
#+END_SRC

** Git

Magit is an awesome interface to git. Summon it with `C-x g`.

#+BEGIN_SRC emacs-lisp
(use-package magit
  :config (setq magit-branch-read-upstream-first 'fallback)
  :bind ("C-x g" . magit-status))
#+END_SRC

Display line changes in gutter based on git history. Enable it everywhere.

#+BEGIN_SRC emacs-lisp
(use-package git-gutter
  :config
  (global-git-gutter-mode 't)
  :diminish git-gutter-mode)
#+END_SRC

TimeMachine lets us step through the history of a file as recorded in git.

#+BEGIN_SRC emacs-lisp
(use-package git-timemachine)
#+END_SRC

** Project management

Still not sure how to make best uses of =projectile=, but everyone and their cow
praises =projectile=.

#+BEGIN_SRC emacs-lisp
(use-package projectile
  :config
  (define-key projectile-mode-map (kbd "s-p") 'projectile-command-map)
  (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)
  (projectile-mode 1))
#+END_SRC

Tell projectile to integrate with =ivy= for completion.

#+BEGIN_SRC emacs-lisp
(setq projectile-completion-system 'ivy)
#+END_SRC

Add some extra completion options via integration with =counsel=. In particular
this enables =C-c p SPC= for smart buffer / file search, and =C-c p s s= for
search via =ag=.

There is no function for projectile-grep, but we could use =counsel-git-grep=
which is similar. Should I bind that to =C-c p s g=?

#+BEGIN_SRC emacs-lisp
(use-package counsel-projectile
  :hook
  (after-init . counsel-projectile-mode))
#+END_SRC


emacs/config.el
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(setq package-enable-at-startup nil)

(package-initialize)
(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(require 'use-package)

(setq use-package-always-ensure t)

(use-package auto-package-update
  :config
  (setq auto-package-update-delete-old-versions t)
  (setq auto-package-update-hide-results t)
  (setq auto-package-update-prompt-before-update t)
  (auto-package-update-maybe))

(use-package benchmark-init
  :config
  ;; To disable collection of benchmark data after init is done.
  (add-hook 'after-init-hook 'benchmark-init/deactivate))

(add-hook 'after-init-hook
          (lambda () (message "loaded in %s" (emacs-init-time))))

(setq gc-cons-threshold 10000000)

;; Restore after startup
(add-hook 'after-init-hook
          (lambda ()
            (setq gc-cons-threshold 1000000)
            (message "gc-cons-threshold restored to %S"
                     gc-cons-threshold)))

(defun find-config ()
  "Edit config.org"
  (interactive)
  (find-file "~/dotfiles/emacs/config.org"))

(global-set-key (kbd "C-c I") 'find-config)

(setq custom-file (make-temp-file "emacs-custom"))

(add-to-list 'load-path "~/.emacs.d/lisp/")

(setq inhibit-startup-screen t)

(setq-default fill-column 80)
(setq-default word-wrap t)
(auto-fill-mode 1)

(tool-bar-mode -1)
(scroll-bar-mode -1)
(menu-bar-mode -1)

(setq ring-bell-function 'ignore)

(defalias 'yes-or-no-p 'y-or-n-p)

(setq backup-directory-alist '(("." . "~/.emacs.d/backup"))
  backup-by-copying t    ; Don't delink hardlinks
  version-control t      ; Use version numbers on backups
  delete-old-versions t  ; Automatically delete excess backups
  kept-new-versions 20   ; how many of the newest versions to keep
  kept-old-versions 5    ; and how many of the old
  )

(setq-default indent-tabs-mode nil)

(setq explicit-shell-file-name "/usr/bin/fish")

(setq dired-listing-switches "-lthG --group-directories-first")

(use-package crux
  :bind (("C-a" . crux-move-beginning-of-line)))

(add-hook 'before-save-hook 'delete-trailing-whitespace)

(use-package smex)

(use-package ivy
    :diminish ivy-mode
    :config
    (ivy-mode t)
    (setq ivy-display-style 'fancy))

(setq ivy-initial-inputs-alist nil)

(use-package counsel
  :bind (("M-x" . counsel-M-x)))

(use-package swiper)
(global-set-key (kbd "C-s") 'counsel-grep-or-swiper)

(use-package ivy-hydra)

(use-package which-key
  :diminish which-key-mode
  :config
  (add-hook 'after-init-hook 'which-key-mode))

(use-package undo-tree
  :defer 5
  :diminish global-undo-tree-mode
  :config
  (global-undo-tree-mode 1))

(use-package avy)

(use-package ace-window
   :config
   (global-set-key (kbd "C-x o") 'ace-window)
   (setq aw-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l)))

(use-package expand-region
  :bind ("C-=" . er/expand-region))

(use-package doom-themes
  :config
  (load-theme 'doom-nord-light t))
(use-package all-the-icons)

(set-frame-font "Source Code Pro 12" nil t)

(use-package feebleline
  :custom
  (feebleline-show-git-branch t)
  (feebleline-show-dir t)
  (feebleline-show-time t)
  (feebleline-show-previous-buffer t)
  :config
  (feebleline-mode 1))

(use-package emojify)

(use-package focus)

(global-hl-line-mode 1)

(use-package smartparens
  :diminish smartparens-mode
  :hook
  (prog-mode . smartparens-mode))

(use-package rainbow-delimiters
  :hook
  (prog-mode . rainbow-delimiters-mode))

(use-package rainbow-mode
  :config
  (setq rainbow-x-colors nil)
  :hook
  (prog-mode . rainbow-mode))

(use-package aggressive-indent)

(add-hook 'prog-mode-hook 'electric-pair-mode)

(use-package smart-dash
  :config
  :hook
  (c-mode . smart-dash-mode)
  (c++-mode . smart-dash-mode))

(add-hook 'prog-mode-hook 'flyspell-prog-mode)

(use-package projectile
  :config
  (projectile-mode))

(setq projectile-completion-system 'ivy)

(use-package counsel-projectile
  :config
  (add-hook 'after-init-hook 'counsel-projectile-mode))

(use-package fzf)

(use-package exec-path-from-shell
  :config
  (exec-path-from-shell-initialize))

(use-package dumb-jump
  :diminish dumb-jump-mode
  :bind (("C-M-g" . dumb-jump-go)
         ("C-M-p" . dumb-jump-back)
         ("C-M-q" . dumb-jump-quick-look)))

(use-package magit
  :bind ("C-x g" . magit-status))

(use-package git-gutter
  :config
  (global-git-gutter-mode 't)
  :diminish git-gutter-mode)

(use-package git-timemachine)

(use-package flycheck
  :config
  (add-hook 'after-init-hook 'global-flycheck-mode)
  (add-to-list 'flycheck-checkers 'proselint)
  (setq-default flycheck-highlighting-mode 'lines)
  ;; Define fringe indicator / warning levels
  (define-fringe-bitmap 'flycheck-fringe-bitmap-ball
    (vector #b00000000
            #b00000000
            #b00000000
            #b00000000
            #b00000000
            #b00000000
            #b00000000
            #b00011100
            #b00111110
            #b00111110
            #b00111110
            #b00011100
            #b00000000
            #b00000000
            #b00000000
            #b00000000
            #b00000000))
  (flycheck-define-error-level 'error
    :severity 2
    :overlay-category 'flycheck-error-overlay
    :fringe-bitmap 'flycheck-fringe-bitmap-ball
    :fringe-face 'flycheck-fringe-error)
  (flycheck-define-error-level 'warning
    :severity 1
    :overlay-category 'flycheck-warning-overlay
    :fringe-bitmap 'flycheck-fringe-bitmap-ball
    :fringe-face 'flycheck-fringe-warning)
  (flycheck-define-error-level 'info
    :severity 0
    :overlay-category 'flycheck-info-overlay
    :fringe-bitmap 'flycheck-fringe-bitmap-ball
    :fringe-face 'flycheck-fringe-info))

(flycheck-define-checker proselint
  "A linter for prose."
  :command ("proselint" source-inplace)
  :error-patterns
  ((warning line-start (file-name) ":" line ":" column ": "
            (id (one-or-more (not (any " "))))
            (message (one-or-more not-newline)
                     (zero-or-more "\n" (any " ") (one-or-more not-newline)))
            line-end))
  :modes (text-mode markdown-mode gfm-mode org-mode))

(use-package company
  :diminish
  :config
  (add-hook 'after-init-hook 'global-company-mode)

  (setq company-idle-delay t)

  (use-package company-anaconda
    :config
    (add-to-list 'company-backends 'company-anaconda)))

(setq company-dabbrev-downcase nil)

(use-package yasnippet
    :diminish yas-minor-mode
    :config
    (add-to-list 'yas-snippet-dirs "~/.emacs.d/yasnippet-snippets")
    (add-to-list 'yas-snippet-dirs "~/.emacs.d/snippets")
    (yas-global-mode)
    (global-set-key (kbd "M-/") 'company-yasnippet))
(use-package yasnippet-snippets)

(use-package markdown-mode
  :commands (markdown-mode gfm-mode)
  :mode (("README\\.md\\'" . gfm-mode)
         ("\\.md\\'" . markdown-mode)
         ("\\.markdown\\'" . markdown-mode))
  :init (setq markdown-command "multimarkdown"))

(use-package haskell-mode)

(use-package hindent)

(use-package ghc
  :config
  (add-hook 'haskell-mode-hook (lambda () (ghc-init))))

(use-package company-ghc
  :config
  (add-to-list 'company-backends 'company-ghc))

;; (use-package anaconda-mode
;;   :config
;;   (add-hook 'python-mode-hook 'anaconda-mode)
;;   (add-hook 'python-mode-hook 'anaconda-eldoc-mode))

(use-package blacken)

(when (executable-find "ipython")
  (setq python-shell-interpreter "ipython"))

(use-package irony
  :hook
  (c-mode . irony-mode)
  (c++-mode . irony-mode))

(use-package company-irony
  :config
  (add-to-list 'company-backends 'company-irony))

(use-package flycheck-irony
  :hook
  (flycheck-mode . flycheck-irony-setup))

(use-package tex
  :demand t
  :ensure auctex
  :config
    (setq-default TeX-engine 'luatex)
    (setq-default TeX-PDF-mode t)
    (setq TeX-view-program-selection '((output-pdf "PDF Tools")))
    (setq reftex-plug-into-AUCTeX t)
    (setq reftex-bibliography-commands '("bibliography" "nobibliography" "addbibresource"))
    (setq reftex-use-external-file-finders t)
    (setq reftex-external-file-finders
          '(("tex" . "kpsewhich -format=.tex %f")
            ("bib" . "kpsewhich -format=.bib %f")))
    (setq TeX-electric-sub-and-superscrip t)
    (setq TeX-electric-math (cons "\\(" "\\)"))
  :hook
    ((LaTeX-mode . visual-line-mode)
     (LaTeX-mode . turn-on-auto-fill)
     (LaTeX-mode . flyspell-mode)
     (LaTeX-mode . LaTeX-math-mode)
     (LaTeX-mode . turn-on-reftex)
     (TeX-after-compilation-finished-functions
       . TeX-revert-document-buffer)))

(use-package auctex-latexmk
  :hook
    (LaTeX-mode . auctex-latexmk-setup))

(use-package company-math
  :config
    (setq company-math-allow-unicode-symbols-in-faces
      '(tex-math font-latex-math-face))
    (setq company-math-allow-latex-symbols-in-faces nil)
    (setq company-math-disallow-unicode-symbols-in-faces nil)
    (add-to-list 'company-backends 'company-math-symbols-unicode)
    (add-to-list 'company-backends 'company-math-symbols-latex))

(setq org-startup-indented 'f)
(setq org-directory "~/org")
(setq org-special-ctrl-a/e 't)
(setq org-default-notes-file (concat org-directory "/notes.org"))
(define-key global-map "\C-cc" 'org-capture)
(setq org-src-window-setup 'current-window)

(use-package org-bullets
  :config
  (setq org-bullets-bullet-list '("∙"))
  (add-hook 'org-mode-hook 'org-bullets-mode))

(setq org-fontify-whole-heading-line t)

(use-package ox-twbs)

(use-package ob-ipython)
(org-babel-do-load-languages
  'org-babel-load-languages
    '((python . t)
      (ipython . t)
      (shell . t)
      (latex . t)
      (C . t)))

(setq org-src-fontify-natively 't)
(setq org-src-tab-acts-natively t)

(org-add-link-type "cite"
     (defun follow-cite (name)
       "Open bibliography and jump to appropriate entry.
        The document must contain \addbibresource{filename} somewhere
        for this to work"
       (find-file-other-window
        (save-excursion
          (beginning-of-buffer)
          (save-match-data
            (re-search-forward "\\\\addbibresource{\\([^}]+\\)}")
            (concat (match-string 1) ".bib"))))
       (beginning-of-buffer)
       (search-forward name))
     (defun export-cite (path desc format)
       "Export [[autocite:cohen93]] as \autocite{cohen93} in LaTeX."
       (if (eq format 'latex)
           (if (or (not desc) (equal 0 (search "autocite:" desc)))
               (format "\\autocite{%s}" path)
             (format "\\autocite[%s]{%s}" desc path)))))

(add-to-list 'org-latex-classes
             '("notes"
               "\\documentclass{article}
   \\usepackage{mathtools}
   \\usepackage{physics}
   \\usepackage{fontspec}
   \\usepackage{unicode-math}
   \\usepackage[style=phys, hyperref=true, backref=true, backend=biber]{biblatex}
   \\usepackage[colorlinks=true, citecolor=red, breaklinks]{hyperref}
   [NO-DEFAULT-PACKAGES]
   [PACKAGES]
                \\setmainfont{Libertinus Serif}
                \\setmathfont{Libertinus Math}
                \\setsansfont{Libertinus Sans}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

(use-package writegood-mode
  :bind ("C-c g" . writegood-mode)
  :config
  (add-to-list 'writegood-weasel-words "actionable"))

(use-package sx
  :config
  (bind-keys :prefix "C-c s"
             :prefix-map my-sx-map
             :prefix-docstring "Global keymap for SX."
             ("q" . sx-tab-all-questions)
             ("i" . sx-inbox)
             ("o" . sx-open-link)
             ("u" . sx-tab-unanswered-my-tags)
             ("a" . sx-ask)
             ("s" . sx-search)))

(use-package notmuch)

(use-package slack
  :commands (slack-start))

(use-package pdf-tools
  :ensure t
  :mode ("\\.pdf\\'" . pdf-view-mode)
  :config
  (pdf-tools-install)
  (setq-default pdf-view-display-size 'fit-page))

(use-package dash)
(use-package dash-functional)

(require 'dired-x)
(setq-default dired-omit-files-p t)
(setq dired-omit-files (concat dired-omit-files "\\|^\\..+$"))

(use-package dired-ranger
  :bind (:map dired-mode-map
              ("W" . dired-ranger-copy)
              ("X" . dired-ranger-move)
              ("Y" . dired-ranger-paste)))
(use-package dired-filter
  :bind (:map dired-mode-map
              ("F" . dired-filter-map)))
(use-package dired-avfs)
(use-package dired-rainbow)
(use-package dired-collapse)
(use-package dired-narrow
  :bind (:map dired-mode-map
              ("/" . dired-narrow)))
(use-package dired-subtree
  :bind (:map dired-mode-map
              ("i" . dired-subtree-insert)
              ("o" . dired-subtree-remove)))

(use-package visual-fill-column
  :config
  (setq visual-fill-column-width 80))
(add-hook 'prog-mode-hook 'turn-on-visual-line-mode)
(add-hook 'text-mode-hook 'turn-on-visual-line-mode)

(use-package fish-completion
  :config
  (global-fish-completion-mode))

(use-package esh-autosuggest
  :hook
  (eshell-mode . esh-autosuggest-mode))
